lom/src/engine/sequenceHandler.go

```diff
@@ -0,0 +1,832 @@
+        }
+        bs, err := GetConfigMgr().GetSequence(data.Action)
+        if err != nil {
+            /* No sequence for this action. Likely stale non-first action or config changed */
+            errCode = LoMMissingSequence
+            errStr = "No sequence found"
+            return
+        }
+        tnow := time.Now().Unix()
+        ctx := make([]*ActionResponseData, 0, len(bs.Actions))
+        ctx = append(ctx, data)
+
+        seq = &sequenceState_t {
+            anomalyInstanceId: anomalyID,
+            sequence: bs,
+            seqStartEpoch: tnow,
+            seqExpEpoch: tnow + int64(bs.Timeout),
+            ctIndex: 1,
+            context: ctx,
+        }
+        p.addSequence(seq, bs.Timeout)
+        return
+    }
+
+    /* reset current request */
+    seq.currentRequest = nil
+
+    /* Move ahead with next action */
+    seq.context = append(seq.context, data)
+    seq.ctIndex++
+
+    /* Save last result */
+    errCode = LoMResponseCode(data.ResultCode)
+    errStr = data.ResultStr
+
+    /* Process the sequence */
+    if data.ResultCode != 0 {
+        /* Need to abort the sequence */
+        return
+    }
+
+    if seq.ctIndex >= len(seq.sequence.Actions) {
+        /* WooHoo we are really complete  */
+        seq.sequenceStatus = sequenceStatus_complete
+        return
+    }
+
+    errCode, errStr, err = p.resumeSequence(seq)
+    if errCode == LoMActionActive {
+        /*
+         * May be a prior seq left behind an active request. We got to wait
+         * till it responds or this sequence timeout, whichever earlier.
+         */
+         errCode = LoMResponseOk
+     }
+    return
+}
+
+/*
+ * resumeSequence
+ *
+ * Validate ctIndex. Attempt to raise request per ctIndex. If failed, return
+ * appropriate error. If succeeded, add it to corresponding client's pending
+ * server requests and list of active requests. Set it as current request of
+ * this sequence.
+ */
+func (p *SeqHandler_t) resumeSequence(seq *sequenceState_t) (errCode LoMResponseCode, errStr string, err error) {
+    regF := GetRegistrations()
+
+    errCode = LoMResponseOk
+    errStr = ""
+    err = nil
+
+    if seq == nil {
+        errCode = LoMSequenceIncorrect
+        errStr = "Resume called on nil sequence"
+        return
+    }
+    if !seq.validate() {
+        errCode = LoMSequenceIncorrect
+        errStr = fmt.Sprintf("Resume called on invalid seq (%v)", *seq)
+        return
+    }
+    if seq.currentRequest != nil {
+        LogInfo("current req is active. Nothing to resume (%v)", *seq)
+        return
+    }
+
+    nextAction := seq.sequence.Actions[seq.ctIndex]
+
+    /* validate action */
+    if regF.GetActiveActionInfo(nextAction.Name) == nil {
+        errCode = LoMActionNotRegistered
+        errStr = fmt.Sprintf("%v", LogError("%s: %s not registered. Abort", seq.sequence.SequenceName,
+                    nextAction.Name))
+        return
+    } else if _, ok := p.activeRequests[nextAction.Name]; ok {
```

Comment:
Is this a redundant check?  In the above code, we are checking "if seq.currentRequest != nil {".    This means there is no active request in this sequence.

Reply:
This action could be active from last seq (timed out).